File: examples/test_motor.c
#include <stdio.h>
#include <math.h>
#include "spikehat.h"

#define PORT_MOTOR   0   /* ポートA: Mアンギュラーモーター */
#define ALIGN_SPEED 10   /* 初期位置合わせの速度 */
#define RUN_SPEED    5   /* motor_start テストの速度 */

/* 現在位置から原点(0度)へ戻す（初期位置合わせ） */
static void return_to_origin(spikehat_t *hat, int port) {
    int cur_pos = 0;
    spikehat_motor_get_position(hat, port, &cur_pos);
    if (cur_pos == 0) return;

    printf("初期位置合わせ: 現在位置 %d 度 -> 0 度\n", cur_pos);
    spikehat_motor_run_for_degrees(hat, port, -cur_pos, ALIGN_SPEED);

    double dur = (fabs((double)cur_pos) / 360.0) / (ALIGN_SPEED * 0.05);
    if (dur < 0.5) dur = 0.5;
    spikehat_sleep(hat, (float)(dur + 0.5));
}

static void print_status(spikehat_t *hat, int port) {
    int speed = 0, pos = 0;
    if (spikehat_motor_get_speed(hat, port, &speed) == 0)
        printf("速度: %d\n", speed);
    if (spikehat_motor_get_position(hat, port, &pos) == 0)
        printf("位置: %d 度\n", pos);
}

int main(void) {
    spikehat_t *hat = spikehat_open(SPIKEHAT_SERIAL_DEFAULT);
    if (!hat) return 1;

    /* ポートA(0)にMアンギュラーモーターを設定 */
    spikehat_port_config(hat, PORT_MOTOR, SPIKEHAT_DEVICE_MOTOR_M);
    spikehat_sleep(hat, 1.0f);

    /* ── run_for_seconds (PID速度制御 + 時間指定) ──────────── */
    printf("=== モーターテスト (run_for_seconds) ===\n");
    return_to_origin(hat, PORT_MOTOR);

    printf("速度5で2秒間回転...\n");
    spikehat_motor_run_for_seconds(hat, PORT_MOTOR, 2.0f, 5);
    spikehat_sleep(hat, 10.0f);
    print_status(hat, PORT_MOTOR);

    printf("速度-3で2秒間回転...\n");
    spikehat_motor_run_for_seconds(hat, PORT_MOTOR, 2.0f, -3);
    spikehat_sleep(hat, 10.0f);
    print_status(hat, PORT_MOTOR);

    spikehat_motor_coast(hat, PORT_MOTOR);

    /* ── motor_start + motor_stop ──────────────────────────── */
    printf("\n=== モーターテスト (motor_start -> motor_stop) ===\n");
    return_to_origin(hat, PORT_MOTOR);

    printf("速度%dでstart, 5秒後にstop...\n", RUN_SPEED);
    spikehat_motor_start(hat, PORT_MOTOR, RUN_SPEED);
    spikehat_sleep(hat, 5.0f);
    spikehat_motor_stop(hat, PORT_MOTOR);
    print_status(hat, PORT_MOTOR);

    /* ── motor_start + motor_coast ─────────────────────────── */
    printf("\n=== モーターテスト (motor_start -> motor_coast) ===\n");
    return_to_origin(hat, PORT_MOTOR);

    printf("速度%dでstart, 5秒後にcoast...\n", RUN_SPEED);
    spikehat_motor_start(hat, PORT_MOTOR, RUN_SPEED);
    spikehat_sleep(hat, 5.0f);
    spikehat_motor_coast(hat, PORT_MOTOR);
    print_status(hat, PORT_MOTOR);

    /* ── motor_pwm + motor_stop ─────────────────────────────── */
    printf("\n=== モーターテスト (motor_pwm -> motor_stop) ===\n");
    return_to_origin(hat, PORT_MOTOR);

    printf("pwm 0.1で回転, 5秒後にstop...\n");
    spikehat_motor_pwm(hat, PORT_MOTOR, 0.1f);
    spikehat_sleep(hat, 5.0f);
    spikehat_motor_stop(hat, PORT_MOTOR);
    print_status(hat, PORT_MOTOR);

    printf("\n完了\n");
    spikehat_close(hat);
    return 0;
}

File: examples/test_motor.py
#!/usr/bin/env python3
"""モーターテスト (Python版)"""
import sys
sys.path.insert(0, '../python')
from spikehat import SpikeHat, DEVICE_MOTOR_M

PORT_MOTOR  = 0   # ポートA: Mアンギュラーモーター
ALIGN_SPEED = 10  # 初期位置合わせの速度
RUN_SPEED   = 5   # motor_start テストの速度


def return_to_origin(hat, port):
    """現在位置から原点(0度)へ戻す（初期位置合わせ）"""
    cur_pos = hat.motor_get_position(port)
    if cur_pos == 0:
        return

    print(f"初期位置合わせ: 現在位置 {cur_pos} 度 -> 0 度")
    hat.motor_run_for_degrees(port, -cur_pos, ALIGN_SPEED)

    dur = (abs(cur_pos) / 360.0) / (ALIGN_SPEED * 0.05)
    if dur < 0.5:
        dur = 0.5
    hat.sleep(dur + 0.5)


def print_status(hat, port):
    try:
        print(f"速度: {hat.motor_get_speed(port)}")
        print(f"位置: {hat.motor_get_position(port)} 度")
    except RuntimeError as e:
        print(f"フィードバックなし: {e}")


with SpikeHat() as hat:
    hat.port_config(PORT_MOTOR, DEVICE_MOTOR_M)
    hat.sleep(1.0)

    # ── run_for_seconds (PID速度制御 + 時間指定) ──────────────
    print("=== モーターテスト (run_for_seconds) ===")
    return_to_origin(hat, PORT_MOTOR)
    
    print("速度5で2秒間回転...")
    hat.motor_run_for_seconds(PORT_MOTOR, 2.0, 5)
    hat.sleep(10.0)
    print_status(hat, PORT_MOTOR)

    print("速度-3で2秒間回転...")
    hat.motor_run_for_seconds(PORT_MOTOR, 2.0, -3)
    hat.sleep(10.0)
    print_status(hat, PORT_MOTOR)
    
    hat.motor_coast(PORT_MOTOR)

    # ── motor_start + motor_stop ──────────────────────────────
    print("\n=== モーターテスト (motor_start -> motor_stop) ===")
    return_to_origin(hat, PORT_MOTOR)

    print(f"速度{RUN_SPEED}でstart, 5秒後にstop...")
    hat.motor_start(PORT_MOTOR, RUN_SPEED)
    hat.sleep(5.0)
    hat.motor_stop(PORT_MOTOR)
    print_status(hat, PORT_MOTOR)

    # ── motor_start + motor_coast ─────────────────────────────
    print("\n=== モーターテスト (motor_start -> motor_coast) ===")
    return_to_origin(hat, PORT_MOTOR)

    print(f"速度{RUN_SPEED}でstart, 5秒後にcoast...")
    hat.motor_start(PORT_MOTOR, RUN_SPEED)
    hat.sleep(5.0)
    hat.motor_coast(PORT_MOTOR)
    print_status(hat, PORT_MOTOR)

    # ── motor_pwm + motor_stop ─────────────────────────────
    print("\n=== モーターテスト (motor_pwm -> motor_stop) ===")
    return_to_origin(hat, PORT_MOTOR)

    print(f"pwm 0.1で回転, 2秒後にstop...")
    hat.motor_pwm(PORT_MOTOR, 0.1)
    hat.sleep(2.0)
    hat.motor_stop(PORT_MOTOR)
    print_status(hat, PORT_MOTOR)

    print("\n完了")
